source/manual/testing.html.md

```diff
@@ -0,0 +1,91 @@
+
+Apps that are automatically deployed all the way to Production should follow the [guidance on healthchecks ("safety checks") in the Continuous Deployment RFC](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md#check-app-is-healthy-ie-it-can-run-in-a-production-environment):
+
+- The healthcheck covers connectivity to all systems the app uses to read or write data e.g. databases, remote file systems, remote caches.
+
+As GOV.UK shifts towards more instantaneous infrastructure, we should be able to run these detailed infrastructure checks as part of pre-merge, end-to-end testing, so that we can learn about configuration issues earlier in the development process.
+
+See also: [how we reuse health check endpoints to alert about transient failures](https://docs.publishing.service.gov.uk/manual/alerts/app-healthcheck-not-ok.html).
+
+## GOV.UK End-to-end Tests
+
+Currently, we have one suite of end-to-end tests: the [Publishing End-to-end Tests](https://github.com/alphagov/publishing-e2e-tests#govuk-publishing-end-to-end-tests). This suite of tests is deprecated and [we have agreed to delete it soon](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md#delete-publishing-e2e-tests). The current tests are slow and brittle and do not run in a realistic GOV.UK environment.
+
+As GOV.UK shifts towards more instantaneous infrastructure, we should be able to create a new suite of tests that accurately represents a real GOV.UK environment. Until then we have agreed to rely on the other forms of testing described here.
+
+## Contract Tests
+
+Also known as "Pact Tests".
+
+Contract tests check that APIs exposed by one app (the "provider") are compatible with other apps that use those APIs ("consumers"). This is done by having a set of shared tests: the tests are run by the provider and the consumers.
+
+Apps that are automatically deployed all the way to Production should follow the [guidance on tests (“safety checks”) in the Continuous Deployment RFC](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md#safety-checks):
+
+- Each endpoint with multiple, internal consumer apps has at least one contract test.
+
+Contract tests are an important part of the overall test strategy for GOV.UK. If we visually add unit and integration tests, the result is a "chain" of test coverage:
+
+```
+Consumer (Unit tests) <---> API (Contract Test) <---> Provider (Unit tests) <---> ...
+```
+
+The chain can be brittle, though: it can't test incremental state changes across multiple apps - think about all the API calls and state changes involved in publishing a document on GOV.UK. We still want higher-level, end-to-end tests to check this.
```

Comment:
You spotted my attempt to blague here 😄.

I'm worried that if I follow your suggestion it will be seen as a contradiction to the earlier statement that it's OK to delete the/some E2E tests (depending on where we end up with the naming).

How about this:

> End-to-end tests are a way of checking for this kind of end-to-end behaviour, but they are also more expensive: they are slower and less robust due to the extensive setup they require to run.

Reply:
That sounds good. I'd perhaps go with "End-to-end tests are an alternative way of checking for this kind of end-to-end behaviour"